nodeadm/internal/containerd/oci_config.go

```diff
@@ -0,0 +1,148 @@
+	"go.uber.org/zap"
+)
+
+type instanceTypeMixin struct {
+	instanceFamilies []string
+	apply            func(*[]byte) error
+}
+
+func (m *instanceTypeMixin) matches(cfg *api.NodeConfig) bool {
+	instanceFamily := strings.Split(cfg.Status.Instance.Type, ".")[0]
+	return slices.Contains(m.instanceFamilies, instanceFamily)
+}
+
+var (
+	// TODO: fetch this list dynamically
+	nvidiaInstances         = []string{"p3", "p3dn", "p4d", "p4de", "p5", "g4", "g4dn", "g5", "g6"}
+	neuronInstances         = []string{"inf1", "inf2", "trn1", "trn1n"}
+	NvidiaInstanceTypeMixin = instanceTypeMixin{
+		instanceFamilies: nvidiaInstances,
+		apply:            applyNvidia,
+	}
+
+	NeuronInstanceTypeMixin = instanceTypeMixin{
+		instanceFamilies: neuronInstances,
+		apply:            applyNeuron,
+	}
+
+	mixins = []instanceTypeMixin{
+		NvidiaInstanceTypeMixin,
+		NeuronInstanceTypeMixin,
+	}
+)
+
+// applyInstanceTypeMixins adds the needed OCI hook options to containerd config.toml
+// based on the instance family
+func applyInstanceTypeMixins(cfg *api.NodeConfig, containerdConfig *[]byte) error {
+	for _, mixin := range mixins {
+		if mixin.matches(cfg) {
+			if err := mixin.apply(containerdConfig); err != nil {
+				return err
+			}
+			return nil
+		}
+	}
+	zap.L().Info("No containerd OCI configuration needed..", zap.String("instanceType", cfg.Status.Instance.Type))
+	return nil
+}
+
+type CommandExecutor interface {
+	CombinedOutput(name string, arg ...string) ([]byte, error)
+}
+
+type RealCommandExecutor struct{}
+
+func (r RealCommandExecutor) CombinedOutput(name string, arg ...string) ([]byte, error) {
+	if isAllowedCommand(name, arg...) {
+		cmd := exec.Command(name, arg...)
+		return cmd.CombinedOutput()
+	}
+	return nil, errors.New("unrecognised command")
+
+}
+
+type FileWriter interface {
+	WriteFileWithDir(filePath string, data []byte, perm fs.FileMode) error
+}
+
+type RealFileWriter struct{}
+
+func (r RealFileWriter) WriteFileWithDir(filePath string, data []byte, perm fs.FileMode) error {
+	return util.WriteFileWithDir(filePath, data, perm)
+}
+
+var execCommand CommandExecutor = RealCommandExecutor{}
+var fileWriter FileWriter = RealFileWriter{}
+var nvidiaCtkCommand = "/usr/bin/nvidia-ctk"
+var nvidiaCtkParams = []string{"--quiet", "runtime", "configure", "--runtime=containerd", "--set-as-default", "--dry-run"}
+
+func isAllowedCommand(command string, args ...string) bool {
+	return command == "/usr/bin/nvidia-ctk" && reflect.DeepEqual(args, nvidiaCtkParams)
+}
+
+// applyNvidia adds the needed Nvidia containerd options using nvidia container toolkit:
+// https://github.com/NVIDIA/nvidia-container-toolkit
+func applyNvidia(containerdConfig *[]byte) error {
+	zap.L().Info("Configuring Nvidia OCI hook..")
+	// before calling nvidia-ctk, we'll write the generated containerd config first
+	// nvidia-ctk will use it as a base
+	err := fileWriter.WriteFileWithDir(containerdConfigFile, *containerdConfig, containerdConfigPerm)
```

Comment:
We need to avoid writing out the config in this file. The mixins should add to the containerdConfig blob and allow the result to be written to disk elsewhere (after all config routines are done).

If that means we stop using nvidia-ctk for this, I think that's okay. All we're doing is adding the same block of TOML with `neuron`/`nvidia` (and the BinaryName) plugged in.

Reply:
I admit it's a bit weird, having to write the containerd file first so that we can generate the resulting conifg and write the final file somewhere else. If you're OK with not using `nvidia-ctk` command to generate the config I can do that. As you said, these are well known values.

Reply:
Yeah I think everything going on in this file would be simpler if we ditched `nvidia-ctk`, let's go that route